src/connectors/s3.rs

```diff
@@ -0,0 +1,373 @@
+    client_receiver: channel::Receiver<S3Client>,
+    bucket: String,
+    client: Option<S3Client>,
+    buffer: Vec<u8>,
+    current_key: String,
+
+    // bookeeping for multipart uploads.
+    upload_id: String,
+    part_number: i32,
+    min_part_size: i32,
+    parts: Vec<CompletedPart>,
+}
+
+#[derive(Deserialize, Debug, Default)]
+pub struct S3Config {
+    aws_access_token: String,
+    aws_secret_access_key: String,
+    aws_region: String,
+    max_clients: Option<usize>,
+    bucket: String,
+    min_part_size: i32,
+    endpoint: Option<String>,
+}
+
+struct S3Meta {
+    key: Option<String>,
+}
+
+impl ConfigImpl for S3Config {}
+
+#[derive(Debug, Default)]
+pub(crate) struct Builder {}
+
+#[async_trait::async_trait]
+impl ConnectorBuilder for Builder {
+    fn connector_type(&self) -> ConnectorType {
+        "s3".into()
+    }
+
+    async fn from_config(
+        &self,
+        _id: &TremorUrl,
+        raw_config: &Option<OpConfig>,
+    ) -> Result<Box<dyn Connector>> {
+        if let Some(config) = raw_config {
+            let mut config = S3Config::new(config)?;
+
+            // Fetch the secrets from the env.
+            config.aws_secret_access_key = env::var(config.aws_secret_access_key)?;
+            config.aws_access_token = env::var(config.aws_access_token)?;
+
+            // Maintain the minimum size of 5 MBs.
+            if config.min_part_size < FIVEMBS {
+                config.min_part_size = FIVEMBS;
+            }
+
+            // Check the validity of given url
+            let endpoint = if let Some(url) = &config.endpoint {
+                let url_parsed = url.parse::<http::Uri>()?;
+                Some(url_parsed)
+            } else {
+                None
+            };
+
+            Ok(Box::new(S3Connector {
+                client_sender: None,
+                config,
+                endpoint,
+            }))
+        } else {
+            Err(ErrorKind::MissingConfiguration(String::from("aws-s3")).into())
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Sink for S3Sink {
+    async fn on_event(
+        &mut self,
+        _input: &str,
+        event: Event,
+        _ctx: &SinkContext,
+        serializer: &mut EventSerializer,
+        _start: u64,
+    ) -> Result<SinkReply> {
+        if self.client.is_none() {
+            if let Ok(new_client) = self.client_receiver.recv().await {
+                self.client = Some(new_client);
+            }
+        } else if let Ok(new_client) = self.client_receiver.try_recv() {
+            self.client = Some(new_client);
+        }
+
+        let ingest_id = event.ingest_ns;
+
+        for (event, meta) in event.value_meta_iter() {
+            // Handle no key in meta.
+            let object_key = match meta.get("key").map(|value| value.to_string()) {
+                Some(key) => key,
+                None => {
+                    self.current_key.clear();
+                    error!("s3: Missing Key Meta Data in Event");
```

Comment:
This will add info about the connector instance to the log:
```suggestion
                    error!("{} Missing Key Meta Data in Event", &ctx);
```